src/ir/lowering/mod.rs

```diff
@@ -412,26 +412,36 @@ trait LowerWhereClauseVec<T> {
 
 impl LowerWhereClauseVec<ir::DomainGoal> for [WhereClause] {
     fn lower(&self, env: &Env) -> Result<Vec<ir::DomainGoal>> {
-        self.iter().map(|wc| wc.lower(env)).collect()
+        self.iter()
+            .flat_map(|wc| match wc.lower(env) {
+                Ok(v) => v.into_iter().map(Ok).collect(),
+                Err(e) => vec![Err(e)],
```

Comment:
mmh can't it be just `=> Err(e)`? (I know there is convertibility from `Vec<Result<T, E>>` to `Result<Vec<T, E>>` but it feels simpler this way in that case)

Reply:
Yeah there's a problem with match arms having different types. If I used `v.into_iter().map(Ok)` then I'd have a `Map` object which of course wouldn't work. So the least common denominator is `Vec`. I may be missing something though.

Reply:
Ah right sorry, ok this is fine the way you wrote I guess.

Reply:
I added `ApplyResult` to make it more ergonomic at least.